### ISSUE TITLE
Trouble Replicating Results

### ISSUE THREAD

**Issue by alexmoore-huma** [imperial-qore/TranAD#33]
I am struggling to reproduce the results you report on the MSL dataset. In the paper you state that you follow other papers and only test your models on the three hardest datasets (A4, C2 and T1). in issue #27 you state:

`This is a refactored version of the code that uses only a single file for each dataset to generate results. We cumulate results over all files of a dataset (mentioned in the paper) to generate the required scores.`

The single dataset you use from MSL in this version of the code is C1, and it produces results very similar to those you have reported in the paper. However, when I update main.py and test the model on C2 it consistently performs very poorly:

{'FN': 0,
 'FP': 1459,
 'Hit@100%': 1.0,
 'Hit@150%': 1.0,
 'NDCG@100%': 1.0000000000000002,
 'NDCG@150%': 1.0000000000000002,
 'ROC/AUC': 0.619258872651357,
 'TN': 457,
 'TP': 135,
 'f1': 0.15615818838352424,
 'precision': 0.0846925967083275,
 'recall': 0.9999999259259313,
 'threshold': 0.018905309658115772}

Considering you report and `AUC = 0.9916` and `F1 = 0.9494` and you are only cumulating across three datasets it doesn't seem like the performance on this dataset is consistent with your results. Can you please confirm that you tested the model on C2 not C1 and if so do you have any explanation as to why the performance is so poor when I run the model locally?

I have a similar question about SMAP, where the results you have reported in the paper appear to be very similar to the results obtained from running the model on P-1 (the file hardcoded into `main.py`) but performance on some of the other datasets is very poor. There are a large number of SMAP files included in the repo can you please confirm which you used to test your model for the paper?

**Comment by shreshthtuli** [comment]
Thanks for raising this. The hyperparameters need to be tuned for each dataset, which are not present in the repo at the moment. We shall share hyperparameter values with each dataset set as soon as they are ready.

**Comment by alexmoore-huma** [comment]
Can I ask about how you tune your hyperparameters? What validation metric are you using? Also how many hyper parameters are you tuning? In the paper you state that you use:

`•  Window size = 10.
• Number of layers in transformer encoders = 1
• Number of layers in feed-forward unit of encoders = 2
• Hidden units in encoder layers = 64
• Dropout in encoders = 0.1`

where are these values used?

**Comment by shreshthtuli** [comment]
We tune the POT threshold parameters in constants.py using grid-search using the F1 score on training set.

**Comment by alexmoore-huma** [comment]
How do you get the f1 score from the training set? My understanding is that there were only labelled anomalies in the test set

**Comment by alexmoore-huma** [comment]
Sorry to pester but I am working on a new anomaly detection model and I want to make sure I am benchmarking its performance correctly. Can you please confirm the following points:

- Did you use every SMAP dataset when aggregating performance (with separate hyper parameter optimisations) or did you use a subset?
- During hyper parameter optimisation you mention using F1 score, how do you calculate this (what are your labels)?
- The SWAT dataset you have provided in the repo is univariate, however, there are multivariate versions of this dataset available online. Are the results you obtained calculated on the univariate or the multivariate case?

**Comment by shreshthtuli** [comment]
1. We used A4 and P1 for SMAP.
2. Sorry for the confusion, we do not have labels for the training set. We use the fraction of positive samples and make sure it is close to the fraction of anomalies in the dataset (given in table 1). 
3. The results in the paper are generated from the multivariate version of SWaT (with 51 dimensions). The repo has a univariate version for simplicity.

**Comment by shreshthtuli** [comment]
We have added params in src/params.json.

**Comment by alexmoore-huma** [comment]
Hello, thanks for uploading the params.json file. Unfortunately, I am still no closer to understanding how you derived these numbers (it seems very odd that different versions of the UCR dataset would require such different POT params). As things stand, without a better explanation of how exactly the params in params.json are derived I am not sure I feel comfortable using them in my research. 

Even when I use these parameters I struggle to replicate the results you reported in your paper. For example across the 4 UCR datasets i get the following averages:

AUC ROC: 0.624
F1: 0.231 
RECALL: 0.250
PRECISION 0.215

The performance you have reported in the paper appears to be the performance from UCR_136 not the average across the 4 datasets (as you claim in the paper). I think the methods section in your paper is very misleading (as this is not the only dataset where your claim to generate results from a different set of examples, see above). I would suggest publishing an amendment or making things much clearer on your readme.

I hope to submit my paper soon but I do want to allow you ample time to respond to these comments as I don't want to misrepresent your work, so please let me know if you have answers to the concerns I have raised.

**Comment by alexmoore-huma** [comment]
Any update on the above?

**Comment by Labaien96** [comment]
Any update on this @alexmoore-huma ??

**Comment by shreshthtuli** [comment]
The results you achieve are very different from the ones reported in the paper. Please ensure:
1. You aggregate true positives, false positives, true negatives and false negatives independetly.
2. You use the same pytorch versions and dataset sizes.
3. You tune hyperparameters for each trace or use the ones provided in the rerpo.
